x/virtualgroup/types/message.go

```diff
@@ -0,0 +1,297 @@
+	return RouterKey
+}
+
+func (msg *MsgDeposit) Type() string {
+	return TypeMsgDeposit
+}
+
+// GetSignBytes implements the LegacyMsg interface.
+func (msg *MsgDeposit) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (msg *MsgDeposit) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromHexUnsafe(msg.FundingAddress)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
+}
+
+// ValidateBasic does a sanity check on the provided data.
+func (msg *MsgDeposit) ValidateBasic() error {
+	_, err := sdk.AccAddressFromHexUnsafe(msg.FundingAddress)
+	if err != nil {
+		return err
+	}
+
+	if !msg.Deposit.IsValid() || !msg.Deposit.Amount.IsPositive() {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "invalid deposit amount")
+	}
+
+	return nil
+}
+
+func NewMsgWithdraw(fundingAddress sdk.AccAddress, globalVirtualGroupID uint32, withdraw sdk.Coin) *MsgWithdraw {
+	return &MsgWithdraw{
+		FundingAddress:       fundingAddress.String(),
+		GlobalVirtualGroupId: globalVirtualGroupID,
+		Withdraw:             withdraw,
+	}
+}
+
+func (msg *MsgWithdraw) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgWithdraw) Type() string {
+	return TypeMsgWithdraw
+}
+
+// GetSignBytes implements the LegacyMsg interface.
+func (msg *MsgWithdraw) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (msg *MsgWithdraw) GetSigners() []sdk.AccAddress {
+	addr, _ := sdk.AccAddressFromHexUnsafe(msg.FundingAddress)
+	return []sdk.AccAddress{addr}
+}
+
+// ValidateBasic does a sanity check on the provided data.
+func (msg *MsgWithdraw) ValidateBasic() error {
+	_, err := sdk.AccAddressFromHexUnsafe(msg.FundingAddress)
+	if err != nil {
+		return err
+	}
+
+	if !msg.Withdraw.IsValid() || msg.Withdraw.Amount.IsNegative() {
+		return errors.Wrap(sdkerrors.ErrInvalidRequest, "invalid deposit amount")
+	}
+	return nil
+}
+
+func NewMsgSwapOut() *MsgSwapOut {
+	return &MsgSwapOut{}
+}
+
+func (msg *MsgSwapOut) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgSwapOut) Type() string {
+	return TypeMsgSwapOut
+}
+
+// GetSignBytes implements the LegacyMsg interface.
+func (msg *MsgSwapOut) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+}
+
+func (msg *MsgSwapOut) GetApprovalBytes() []byte {
+	fakeMsg := proto.Clone(msg).(*MsgSwapOut)
+	fakeMsg.SuccessorSpApproval.Sig = nil
+	return fakeMsg.GetSignBytes()
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (msg *MsgSwapOut) GetSigners() []sdk.AccAddress {
+	addr, _ := sdk.AccAddressFromHexUnsafe(msg.OperatorAddress)
+	return []sdk.AccAddress{addr}
+}
+
+// ValidateBasic does a sanity check on the provided data.
+func (msg *MsgSwapOut) ValidateBasic() error {
+	return nil
```

Comment:
Need to validate some fields, e.g., `global_virtual_group_ids`

Reply:
Fixed